[KB] Update PR/MR comments troubleshooting to reflect current advice (#2067)
semgrep/semgrep-docs: docs/kb/semgrep-appsec-platform/missing-pr-comments.md

Creating git worktree from '104...950' to scan baseline.
...
┌─────────────┐
│ Scan Status │
└─────────────┘
  Scanning 4 files tracked by git with 2 Code rules, 34 Supply Chain rules:
```

The initial scan, which occurs at the current commit for the pull request `fcc...d21`, scans 52 files and identifies 60 findings. The baseline scan, which occurs at `104...950`, scans 4 files. Baseline scans typically scan fewer files than the original scan, as they only need to scan files and rules that have findings in the initial scan to determine which of those findings were present before the changes made in the pull or merge request.

If you review the scans that are not generating comments and find that they are not diff-aware, and you have followed the preceding guidance, feel free to [reach out to Semgrep support](/docs/support) for help.

## Have you correctly configured your policies?

### Code: The rule is in Comment or Block mode

To receive comments for a Code rule, the rule must be in the [Comment or Block policy mode](https://semgrep.dev/docs/semgrep-code/policies#block-a-pr-or-mr-through-rule-modes). Rules in Monitor mode do not generate comments.

> ### Secrets: The Rule is in Comment or Block mode, and the secret is valid

Author: ```suggestion
### Secrets: The rule is in Comment or Block mode, and the secret is valid
```

> ### Secrets: The rule is in Comment or Block mode, and the secret is valid

Author: ```suggestion
### Secrets: The rule is in Comment or Block mode, and validation settings match
```

![Semgrep GitHub app permissions: all repositories](/img/gh-app-permissions-all.png)
**Figure** Permissions for all repositories.

![Semgrep GitHub app permissions - select repositories](/img/gh-app-permissions-select.png)
**Figure** Permissions for select repositories. Ensure the repositories you have onboarded to Semgrep AppSec Platform are included here.

### Azure DevOps, GitLab, and Bitbucket

For MR comments on GitLab and PR comments on Bitbucket or Azure DevOps, ensure that you have [set up a connection to your SCM organization or project](/deployment/connect-scm), including a token with appropriate permissions to comment on MRs or PRs.

#### Azure DevOps

See [Enable Azure pull request comments](/docs/semgrep-appsec-platform/azure-pr-comments) for token and configuration guidelines.

#### GitLab

The GitLab token should have `api` scope. See [Enable GitLab merge request comments](/docs/semgrep-appsec-platform/gitlab-mr-comments) for details.

Previously, it was possible to supply an `api` scoped token via the GitLab CI/CD pipeline, rather than through the SCM connection, but this is no longer supported.

#### Bitbucket

> The Bitbucket token should be a repository access token (or workspace access token, for Bitbucket Cloud Premium only). See [Enable Bitbucket pull request comments](/category/bitbucket-pr-comments) for details.

Author: almost, but there's a typo in the suggestion 😂 I'll come back to this one

> The Bitbucket token should be a repository access token (or workspace access token, for Bitbucket Cloud Premium only). See [Enable Bitbucket pull request comments](/category/bitbucket-pr-comments) for details.

Author: ```suggestion
The Bitbucket token must be a repository access token or a workspace access token. See [Enable Bitbucket pull request comments](/category/bitbucket-pr-comments) for details.
```

> ## Are you running diff-aware scans?

Reviewer: actually, after looking through the whole thing, I'd reframe that suggestion as -- what if we put first-seen first, then policies, then leave the rest as they are (tokens, then diff-aware). I don't actually have the data for this but I suspect sorting short-to-long section and sorting common-to-uncommon problem get us roughly the same results on this page.

> ## Are you running diff-aware scans?

Author: I'd rather not reorganize the whole doc right now. At the moment, it's set up to go from "most fundamental" to "most niche", so there is logic to the order, but I'd be happy to review another PR to change that later on if there is a better order.

For Managed Scans, Semgrep always runs diff-aware scans on pull request events. (Standard full scans are run at scheduled intervals.)

> For GitHub Actions and GitLab CI/CD, if you are using the recommended configuration, Semgrep typically runs diff-aware scans on pull or merge requests by default.

Author: Hmm I may come back to this one too, I'm not sure I like my combined use of 'typically' and 'by default' on reflection.

In other SCMs or CI systems, or with unusual pipeline configurations: diff-aware scans may require additional setup. Review the [configuration instructions](/category/pr-or-mr-comments) for your SCM or [custom configuration for your CI jobs](/deployment/customize-ci-jobs#set-up-diff-aware-scans) and ensure you have configured your scans correctly.

### Identify a diff-aware scan

Semgrep diff-aware scans are most easily identified by reviewing three items in the scan log:

* The triggering event
* The number of files scanned
* Whether a baseline scan is conducted

> 1. The triggering event for the scan in GitHub or GitLab should typically be `pull_request`. This is the easiest to find but the least reliable, since it's possible to configure diff-aware scans on other event types. In the scan log, this appears as:

Author: In what sense is it confusing?

The reason I didn't use subheading is that to me it feels weird to list out the three things and then also immediately have the first one appear as a header, but that might just be me, or maybe there's a better way than to list them all like that if they're headers?

> 1. The triggering event for the scan in GitHub or GitLab should typically be `pull_request`. This is the easiest to find but the least reliable, since it's possible to configure diff-aware scans on other event types. In the scan log, this appears as:

Author: fixed this up (and added a quick note about skipped baselines)

```

This would most likely be a diff-aware scan, unless you are doing a test on a very small repository.

However, if you instead see something like:

```
Scanning 1002 files tracked by git with 1971 Code rules, 858 Secrets rules, 3619 Supply Chain rules:
```

and the repository's total number of files is around 1000, then this most likely is not a diff-aware scan.

3. Finally, during the process of a diff-aware scan, Semgrep actually conducts two scans: one at the current tip or head of the PR, and one at the baseline ref or commit.

The following log is anonymized and truncated for clarity, and the exact format of the log may evolve over time. It shows the key item to review, the two distinct `Scan Status` entries:

```
┌────────────────┐
│ Debugging Info │
└────────────────┘

  SCAN ENVIRONMENT
  versions    - semgrep 1.76.0 on python 3.11.9
  environment - running in environment github-actions, triggering event is pull_request
Fixing git state for github action pull request
Not on head ref: fcc...d21; checking that out now.

  CONNECTION
Using 104...950 as the merge-base of f5e...1a7 and fcc...d21

>   Initializing scan (deployment=testdep, scan_id=29062823)

Author: The idea isn't for it to necessarily be real or fake, just to look like a test example. I'm not sure I understand the concern about it potentially being real, anyhow?
